### PR TITLE
Make `#![feature(bind_by_move_pattern_guards)]` sound without `#[feature(nll)]`

### LINKED ISSUE #27282: Can mutate in match-arm using a closure
**Update from pnkfelix**: Note that this is now fixed by NLL (see #50783). This bug just remains open because @pnkfelix thinks our policy is to keep NLL-fixed-by-NLL bugs open until we make NLL the default for `rustc`.

----

## STR

``` Rust
fn main() {
    match Some(&4) {
        None => {},
        ref mut foo
            if {
                (|| { let bar = foo; bar.take() })();
                false
            } => {},
        Some(s) => println!("{}", *s)
    }
}
```
## Actual Results

```
playpen: application terminated abnormally with signal 4 (Illegal instruction)
```

### LINKED ISSUE #31287: Pattern guard can consume value that is being matched
Following compiles, and crashes when trying to print moved from value:

``` rust
fn main() {
    let a = Some("...".to_owned());
    let b = match a {
        Some(_) if { drop(a); false } => None,
        x                             => x,
    };
    println!("{:?}", b);
}
```

### PR DESCRIPTION
**Pull request by Centril**
Implements https://github.com/rust-lang/rust/issues/15287#issuecomment-507054617 making `#![feature(bind_by_move_pattern_guards)]]` sound without also having `#![feature(nll)]`. The logic here is that if we see a `match` guard, we will refuse to downgrade NLL errors to warnings. This is in preparation for hopefully stabilizing the former feature in https://github.com/rust-lang/rust/pull/63118.

As fall out from the implementation we also:
Fixes https://github.com/rust-lang/rust/issues/31287
Fixes https://github.com/rust-lang/rust/issues/27282

r? @matthewjasper

### CODE REVIEW

**Review by matthewjasper — COMMENTED** [review]
This mostly looks good, but I wasn't expecting all matches with guards to result in hard borrowck errors. I'm not sure if @rust-lang/lang has an opinion here.

File: src/librustc_mir/hair/pattern/check_match.rs [hunk 1]
```diff
@@ -130,26 +132,27 @@ impl<'a, 'tcx> PatternContext<'a, 'tcx> {
...
+        source: hir::MatchSource
+    ) {
         for arm in arms {
             // First, check legality of move bindings.
             self.check_patterns(arm.guard.is_some(), &arm.pats);
 
             // Second, if there is a guard on each arm, make sure it isn't
             // assigning or borrowing anything mutably.
             if let Some(ref guard) = arm.guard {
+                self.signalled_error = SignalledError::SawSomeError;
```
[the remark this answers is not in the document]
Comment (Centril) [reply src/librustc_mir/hair/pattern/check_match.rs]: Fixed. :)

**Comment by Centril** [comment]
> [...] but I wasn't expecting all matches with guards to result in hard borrowck errors. [...]

Fwiw, I tried some hacks to avoid it but that went nowhere... =)

**Comment by Centril** [comment]
@bors r-

Failed in https://github.com/rust-lang/rust/pull/63120#issuecomment-516244972.
Looks like cargotest is testing servo which uses cssparser 0.24.0 which in turn has a bug.

**Comment by Centril** [comment]
Rebased & bumped cargotests ref of servo to `9043f247d9b031ed285e880e4b90aa523d4a63ae` (latest at the time of writing) which has a version of cssparser without the bug.

@bors r=matthewjasper rollup=never

**Comment by Centril** [comment]
Failed in https://github.com/rust-lang/rust/pull/63182#issuecomment-517158575, @bors r-
Raw log: https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/4878/logs/1049

@SimonSapin @Manishearth Can you please advise as to how I can fix this? Servo is causing problems in cargotest.

**Comment by Manishearth** [comment]
No idea what's going on there

**Comment by Centril** [comment]
@bors retry rollup=never p=1

Let's isolate this PR to see if we can reproduce the error when it is tested alone.

**Comment by SimonSapin** [comment]
The likely-most-relevant part of the log is:

```
2019-08-01T07:27:27.9889662Z error: linker `lld-link.exe` not found
2019-08-01T07:27:27.9902164Z   |
2019-08-01T07:27:27.9902349Z   = note: The system cannot find the file specified. (os error 2)
2019-08-01T07:27:27.9902736Z 
2019-08-01T07:27:27.9903043Z note: the msvc targets depend on the msvc linker but `link.exe` was not found
2019-08-01T07:27:27.9903384Z 
2019-08-01T07:27:27.9903507Z note: please ensure that VS 2013, VS 2015, VS 2017 or VS 2019 was installed with the Visual C++ option
2019-08-01T07:27:27.9903587Z 
2019-08-01T07:27:28.0078423Z error: aborting due to previous error
```

… which appears unrelated to this PR. But then it also appears unrelated to other PRs in that rollup.

This is a long way of saying I also have no idea what’s going on here.

**Comment by Centril** [comment]
Same error again, but this time tested standalone and not in a rollup:
```
   Compiling proc-macro2 v0.4.26
   Compiling autocfg v0.1.4
error: linker `lld-link.exe` not found
  |
  = note: The system cannot find the file specified. (os error 2)

note: the msvc targets depend on the msvc linker but `link.exe` was not found

note: please ensure that VS 2013, VS 2015, VS 2017 or VS 2019 was installed with the Visual C++ option

error: aborting due to previous error

error: Could not compile `proc-macro2`.
warning: build failed, waiting for other jobs to finish...
error: build failed
thread 'main' panicked at 'tests failed for https://github.com/servo/servo', src\tools\cargotest\main.rs:86:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```
cc @alexcrichton @pietroalbini

**Comment by pietroalbini** [comment]
The Servo failure is probably caused by https://github.com/servo/servo/pull/23256. The last Servo commit that didn't use ldd is https://github.com/servo/servo/commit/caac107ae8145ef2fd20365e2b8fadaf09c2eb3b, so bumping to that should work. cc @jdm

**Comment by Centril** [comment]
Updated the servo ref to `caac107ae8145ef2fd20365e2b8fadaf09c2eb3b` per ^---.

@bors r=matthewjasper

**Comment by pietroalbini** [comment]
@bors retry

Yielded priority to https://github.com/rust-lang/rust/pull/63242

**Pull request merged**